### PR TITLE
autoconfig: WSC: Change config data network key to be dynamic

### LINKED ISSUE #449: [BUG] <tlvf>: <M2 is using a fixed-length password>
**Describe the bug** -
Currently, test 5.6.2 keeps failing when using Marvell as CTT Agent.

The reason for this is that we send the M2 with a fixed-length attribute
for the password. In cases that the actual password is shorter than the maximum, 
the agent may use zeroes at the end.

The solution is to make the sWscAttrNetworkKey attribute to be a dynamic size, to prevent an agent from adding zeroes at the end.

### PR DESCRIPTION
**Pull request by CoralMalachi**
Currently, test 5.6.2 keeps failing when using Marvell as CTT Agent.

The reason for this is that we send the M2 with a fixed-length attribute
for the password. In cases that the actual password is shorter than the maximum,
the agent may use zeroes at the end.

In order to solve this, the network_key attribute is added manually to cConfigData in the WSC_Attributes YAML file.
Also, sWscAttrNetworkKey is removed since it is not needed anymore.

Fixes #449 

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

### CODE REVIEW

File: framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml
Comment (tomereli) [inline framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml]: extra line break

File: framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml
Comment (tomereli) [inline framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml]: extra line break

File: framework/tlvf/test/tlvf_test.cpp
Comment (tomereli) [inline framework/tlvf/test/tlvf_test.cpp:219]: This should be done in a separate commit which sets a short password and compares it in the parse, making the test fail.

**Review by tomereli — APPROVED** [review]

File: framework/tlvf/test/tlvf_test.cpp
Comment (morantr) [inline framework/tlvf/test/tlvf_test.cpp]: This whole commit is basically wrong. On afb6e383789d21e0d63e6cb6b60e40a1deb1f9c1 you changed the yaml so the string will need to be dynamically allocated, and you copy it as if it was constant array size. This will end up with a segmentation fault, so you cannot push it (according to the commit guidelines: https://github.com/prplfoundation/prplMesh/blob/master/CONTRIBUTING.md#commits)

But even if weren't change the yaml file to use dynamic list, it was still redundant to convert the char* to string and compare the strings at the end, since it will always be identical strings.

EDIT: ok I understand now that you change the way you set the string on the next commit so it will not break after all. But still it has redundant parts, and I don't see the point of doing it if in the next commit to change it.

File: framework/tlvf/test/tlvf_test.cpp
Comment (morantr) [inline framework/tlvf/test/tlvf_test.cpp]: This test is basically testing the tlv_factory simple list and is already tested on a simple list test, which makes this check unnecessary.

**Review by morantr — CHANGES_REQUESTED** [review]

**Comment by morantr** [comment]
Regarding commit 2bcfebe2b9884d00a50daa009889fe61659687f1, I think it might cause a problem since now the getter of the string data that return char* will return an un-null-terminated string and this could cause some issues. I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.

File: framework/tlvf/test/tlvf_test.cpp
Comment (tomereli) [reply framework/tlvf/test/tlvf_test.cpp]: Actually this commit is based on @arnout 's suggestion to include one commit which reproduces the issue, then add a commit which fixes it. It came with the context of adding tests which fails and only then implement the fix, it should help for better review. Since @arnout is OOO this week, we can't wait for him to provide feedback, so your call - if you still think its better to remove this commit, I will.

File: framework/tlvf/test/tlvf_test.cpp
Comment (tomereli) [reply framework/tlvf/test/tlvf_test.cpp]: ok, will be removed.

**Comment by tomereli** [comment]
> Regarding commit [2bcfebe](https://github.com/prplfoundation/prplMesh/commit/2bcfebe2b9884d00a50daa009889fe61659687f1), I think it might cause a problem since now the getter of the string data that return char* will return an un-null-terminated string and this could cause some issues. I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.

It is the wrong usage by the caller, the function returns `char *` so the caller should not use it for string operations. This is why we have another API which returns `std::string` - see `tlvWscM2.h: std::string manufacturer_str()` for example.

File: framework/tlvf/test/tlvf_test.cpp
Comment (tomereli) [reply framework/tlvf/test/tlvf_test.cpp]: Edit - I removed it

**Review by morantr — APPROVED** [review]

**Comment by morantr** [comment]
> > Regarding commit [2bcfebe](https://github.com/prplfoundation/prplMesh/commit/2bcfebe2b9884d00a50daa009889fe61659687f1), I think it might cause a problem since now the getter of the string data that return char* will return an un-null-terminated string and this could cause some issues. I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.
> 
> It is the wrong usage by the caller, the function returns `char *` so the caller should not use it for string operations. This is why we have another API which returns `std::string` - see `tlvWscM2.h: std::string manufacturer_str()` for example.

I approved it but I still think that leaving it as it is dangerous.

**Comment by tomereli** [comment]
> > > Regarding commit [2bcfebe](https://github.com/prplfoundation/prplMesh/commit/2bcfebe2b9884d00a50daa009889fe61659687f1), I think it might cause a problem since now the getter of the string data that return char* will return an un-null-terminated string and this could cause some issues. I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.
> > 
> > 
> > It is the wrong usage by the caller, the function returns `char *` so the caller should not use it for string operations. This is why we have another API which returns `std::string` - see `tlvWscM2.h: std::string manufacturer_str()` for example.
> 
> I approved it but I still think that leaving it as it is dangerous.

It is dangerous but nothing we can do about it - it is the only way to allow users to modify list members, in this case the type is `char` - if someone tries to use this in a print function it is a BUG, this is why we have the `_str()` version.

**Comment by morantr** [comment]
> > > > Regarding commit [2bcfebe](https://github.com/prplfoundation/prplMesh/commit/2bcfebe2b9884d00a50daa009889fe61659687f1), I think it might cause a problem since now the getter of the string data that return char* will return an un-null-terminated string and this could cause some issues. I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.
> > > 
> > > 
> > > It is the wrong usage by the caller, the function returns `char *` so the caller should not use it for string operations. This is why we have another API which returns `std::string` - see `tlvWscM2.h: std::string manufacturer_str()` for example.
> > 
> > 
> > I approved it but I still think that leaving it as it is dangerous.
> 
> It is dangerous but nothing we can do about it - it is the only way to allow users to modify list members, in this case the type is `char` - if someone tries to use this in a print function it is a BUG, this is why we have the `_str()` version.

You used it in print function in this PR. I have suggested a solution.

**Comment by tomereli** [comment]
> You used it in print function in this PR. I have suggested a solution.

Where?

**Comment by morantr** [comment]
>I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.

**Comment by morantr** [comment]
> > You used it in print function in this PR. I have suggested a solution.
> 
> Where?

If you asked about my solution then:
>I think you should modify this function as well to call to the function that returns the string object that will be successfully created, and return its "c_str()" return value.

If you asked about where is the use in this PR then:
d8df0f9208bcaaa7c50c3b667c9325d87f000981 framework/tlvf/test/tlvf_test.cpp line 228

**Pull request merged**
